### PR TITLE
[1.11.2] Added support for Village capabilities

### PR DESCRIPTION
**Pull request by Madsthunder**
This PR adds the fields and methods required for Capabilities into the Village class, similar to Entity, TileEntity, ItemStack, and World.

### CODE REVIEW

**Comment by shadowfacts** [comment]
Why is this necessary/useful?

**Comment by mezz** [comment]
I also think this seems very strange. What are the village capabilities used for?

**Comment by Madsthunder** [comment]
@shadowfacts This is useful because it allows a modder to store extra data about a village. For example, if a modder needs a unique identifier for a village they can do so by creating a capability that serializes/deserializes a UUID. This would remove the need for modders to try to determine which village a villager is from by assuming from NBT, because the village would have a unique identifier.

**Comment by Madsthunder** [comment]
@mezz The village capabilities could be used for multiple things, like if a modder wanted to add economy into the game, the village could represent a nation, the modder could then add a capability called Wealth or something along those lines  and assign it to the appropriate Village.

File: patches/minecraft/net/minecraft/village/Village.java.patch [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
++        this.capabilities = net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(this);
+     }
+ 
+     public void func_82691_a(World p_82691_1_)
+@@ -466,6 +468,8 @@
+                 this.field_82693_j.put(nbttagcompound1.func_74779_i("Name"), Integer.valueOf(nbttagcompound1.func_74762_e("S")));
+             }
+         }
++        if (this.capabilities != null && p_82690_1_.func_74764_b("ForgeCaps")) this.capabilities.deserializeNBT(p_82690_1_.func_74775_l("ForgeCaps"));
++
```
Comment (mezz) [inline patches/minecraft/net/minecraft/village/Village.java.patch]: please remove extra newline here

File: patches/minecraft/net/minecraft/village/Village.java.patch [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+-public class Village
++public class Village implements net.minecraftforge.common.capabilities.ICapabilitySerializable<NBTTagCompound>
+ {
+     private World field_75586_a;
+     private final List<VillageDoorInfo> field_75584_b = Lists.<VillageDoorInfo>newArrayList();
+@@ -43,11 +43,13 @@
+ 
+     public Village()
+     {
++        this(null);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/village/Village.java.patch]: This seems slightly questionable? May be better to just set the capabilities here too. I'm open to other opinions though.

File: patches/minecraft/net/minecraft/village/Village.java.patch [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+         p_82689_1_.func_74782_a("Players", nbttaglist1);
++        if (this.capabilities != null) p_82689_1_.func_74782_a("ForgeCaps", this.capabilities.serializeNBT());
+     }
+ 
+     public void func_82692_h()
+@@ -553,4 +558,29 @@
+             this.field_75590_b = p_i1674_3_;
+         }
+     }
++
```
Comment (mezz) [inline patches/minecraft/net/minecraft/village/Village.java.patch]: Could use a 
`/* ======================================== FORGE START =====================================*/`
section for good measure

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by Madsthunder** [comment]
Changed the constructor call to Village(World) in Village() to a gatherCapabilites call, removed the extra line at the end of writeVillageToNBT and added a Forge Start/End surrounding the new methods.

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Can you make a small forge debug mod that uses this capability, to ensure it is working as intended and nothing else is needed?
There are several in `net.minecraftforge.debug`, if you need any help getting set up there is a [tutorial video here](https://www.youtube.com/watch?v=pLWQk6ed56Q).

**Comment by Madsthunder** [comment]
@mezz Would you mind if I integrated it into the existing TestCapabilityMod?

**Comment by mezz** [comment]
Sure, as long as it's not interacting with the existing stuff there it should be fine.

**Comment by LatvianModder** [comment]
Forge should figure out a way to store unique IDs for - Worlds/dimensions, villages and the server...

**Comment by Madsthunder** [comment]
@mezz Fixed the AttachCapabilitiesEvent not firing correctly and added Villages to the TestCapabilityMod. The mod also checks every tick for players entering and exiting villages and prints it to the chat.

**EDIT:**
One thing to note is that this only works on server for some reason

**Comment by rguy96** [comment]
@mezz Any idea when the next batch of PRs are being pulled?

**Comment by mezz** [comment]
@rguy96 Probably after PAX South, so in February.

**Comment by LexManos** [comment]
between moving, and PAX, I'm tied up till the 3rd. Also this one needs to be looked into quite a but. I'm not a particular fan of just sticking caps on every possible thing. I do think this needs a few more concrete explained if its use case.

**Comment by mezz** [comment]
@Madsthunder can you address Lex's comment by giving a few more concrete examples of uses for this event?

**Comment by Madsthunder** [comment]
Other than the examples above, I can't really think of any off the top of my head, and the use cases I have for it are a bit too complex to put into one comment...

**Comment by mezz** [comment]
@Madsthunder can you update this to resolve conflicts?

I think this is a good way to attach extra data to villages, which is not simple to do otherwise.
I'm not exactly sure capabilities are the perfect fit because I can't easily imagine how mods would share the capabilities. However it's not a very invasive patch and the implementation is simple.
I think it's up to Lex to look into this a bit more when he has more time to explain.

File: src/test/java/net/minecraftforge/test/TestCapabilityMod.java [hunk 4]
```diff
@@ -186,7 +190,51 @@ public void attachTileEntity(AttachCapabilitiesEvent<TileEntity> event)
...
+        if(!event.world.isRemote)
+        {
+            List<EntityPlayer> players = event.world.playerEntities;
+            int i = 0;
+            for(Village village : event.world.villageCollectionObj.getVillageList())
+            {
+                if(village.hasCapability(TEST_CAP, null))
+                {
+                    boolean playerInRadius = false;
+                    for(EntityPlayer player : players)
```
Comment (mezz) [inline src/test/java/net/minecraftforge/test/TestCapabilityMod.java:214]: this `for` loop needs braces

File: patches/minecraft/net/minecraft/village/Village.java.patch [hunk 5]
```diff
@@ -135,3 +170,36 @@
...
+@@ -553,4 +574,32 @@
+             this.field_75590_b = p_i1674_3_;
+         }
+     }
++
++    /* ======================================== FORGE START =====================================*/
++    private net.minecraftforge.common.capabilities.CapabilityDispatcher capabilities;
++    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
++    {
++        if (getCapability(capability, facing) != null)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/village/Village.java.patch]: `hasCapability` is meant to be a faster cheaper check than `getCapability` so it should not call `getCapability`.

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/village/Village.java.patch [hunk 6]
```diff
@@ -24,19 +45,20 @@
...
  
 +    @Deprecated //Hasn't worked since 1.9, use UUID version below.
      public int func_82684_a(String p_82684_1_)
      {
+-        Integer integer = (Integer)this.field_82693_j.get(p_82684_1_);
 +        return this.getPlayerReputation(findUUID(p_82684_1_));
 +    }
 +
-+    public int getPlayerReputation(UUID p_82684_1_)
++    public int getPlayerReputation(UUID playerName)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/village/Village.java.patch]: Dunno how you messed this up, but fix it.

**Review by mezz — APPROVED** [review]

**Pull request merged**